2097 CoordConv transform

marksgraham opened the pull request:
Fixes #2097  .

### Description
Implements a transformation that appends spatial coordinates to the input as new channels, inspired by the  [CoordConv paper](https://arxiv.org/abs/1807.03247). Have found this improves the performance of my segmentation networks.

### Status
**Ready**

### Types of changes
<!--- Put an `x` in all the boxes that apply, and remove the not applicable items -->
- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`. **I get a pytype error that I can't resolve - help appreciated**
- [x] Quick tests passed locally by running `./runtests.sh --quick --unittests`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

wyli commented:
thanks, for the typing error, a simple workaround would be removing ` -> Union[np.ndarray, torch.Tensor]` in the newly added classes

rijobro commented on monai/transforms/spatial/array.py:
```diff
@@ -1700,3 +1701,50 @@ def __call__(
+        coord_channels = np.ones((len(self.spatial_channels), *spatial_dims)).astype(img.dtype)
+
+        for i, dim in enumerate(self.spatial_channels):
+            ones = np.ones((1, *spatial_dims))
```
Could the calculation of the coordinates be done as:

```python
shape = img.shape[1:]
num_channels = img.shape[0]
coords = sum(np.ogrid[tuple(slice(0, i) for i in shape)])
coords = np.repeat(coords[None], repeat=num_channels, axis=0)
```

rijobro replied:
You could remove the `for s in spatial_channels` by requiring use input to be 0-based on the spatial dimensions (e.g., `(0,1,2)` for `(x,y,z)`).

rijobro commented on monai/transforms/spatial/array.py:
```diff
@@ -1700,3 +1701,51 @@ def __call__(
+
+        spatial_dims = img.shape[1:]
+        num_spatial_dims = len(spatial_dims)
+        coord_channels = np.mgrid[tuple(slice(0, i) for i in spatial_dims)]
+        # normalise so coordinates lie between [-0.5,0.5]
+        normalisation = np.expand_dims(np.array(spatial_dims) - 1, axis=tuple(np.arange(1, num_spatial_dims + 1)))
+        coord_channels = coord_channels / normalisation - 0.5
```
if you want results between -0.5 and 0.5, you could use something meshgrid?

```suggestion
        coord_channels = np.array(np.meshgrid(*tuple(np.linspace(-0.5, 0.5, s) for s in spatial_dims)))
```

marksgraham commented:
Thanks for comments. I have:
- Fixed the typing error (I think!)
- Used @rijobro's cleaner code for generating the channels. I've decided to stick with the `spatial_channels` argument starting from 1 rather than 0 but can change if you feel stronlg they should start from 0. I looked for some precedent in other transforms but couldn't find any.
- Finally, I've renamed the transform to `AddCoordinateChannels`, and provided a bit of motivation for when you might want to use it in the docs. It isn't correct to call it CoordConv because CoordConv changes the convolution layer itself to append spatial channels for each convolution, while my code only appends these channels to the input data as a transform. I think my version has particular use for medical imaging examples where e.g. you might train on patches of 3D volumes that are all aligned to a common space. Appending the coordinate frame to the patch might allow the network to understand where the patch is in the image - CoordConv only appends the local reference coordinates of the input feature map.

rijobro commented:
Thanks, looks great. Since the output is a concatenation of the channels and the original image, in the unit test it might be worth checking that the first part of the image matches the original image?

I suggested another update to the calculation of the coords (sorry for playing a bit of code golf!) if you want to add that too.

marksgraham commented:
Thanks - I've taken your suggestion for meshgrid (had to add `indexing='ij'` to get the first two dimensions right), and added that unit test.

rijobro commented:
Looks good to me, happy to merge once the checks are passing!

wyli reviewed (approved):
thanks! @marksgraham @rijobro
